File: route/rides.js
import dotenv from 'dotenv';
import got from 'got';
import express from 'express';
import RideData from '../model/ride_data.js';
import RideRecord from '../model/ride_record.js';
import { findRideRecordByID, findRideDataByRecordID } from './middleware.js';

dotenv.config();
const router = express.Router();

// Get all
router.get('/', async (req, res) => {
    try {
        const rideRecord = await RideRecord.find().sort({ createdAt: -1 }).limit(20);

        res.json(rideRecord);
    } catch (err) {
        res.status(500).json({ message: err.message });
    }
});

// Get one
router.get('/:rideObjID', [findRideRecordByID, findRideDataByRecordID], async (req, res) => {
    try{    
        res.json({
            _id: res.rideRecord._id,
            rideName: res.rideRecord.rideName,
            rideDate: res.rideRecord.rideDate,
            rideScore: res.rideRecord.rideScore,
            rideMeta: res.rideRecord.rideMeta,
            rideData: res.rideData
        });
    } catch (err) {
        return res.status(500).json({ 
            message: err.message 
        });
    }
});

// Create one
router.post('/', async (req, res) => {
    let rideRecord;
    let allRideData = [];

    try {
        // Create new rideRecord document
        rideRecord = new RideRecord({
            rideName: req.body.rideName,
            rideDate: req.body.rideDate
        });

        // Create documents for rideData belonging to rideRecord
        for (let k in req.body.rideData) {
            allRideData.push(new RideData({
                timestamp: req.body.rideData[k].timestamp,
                metadata: { rideRecordID: rideRecord._id },
                locationLat: req.body.rideData[k].locationLat,
                locationLong: req.body.rideData[k].locationLong,
                accelerometerZ: req.body.rideData[k].accelerometerX,
                accelerometerX: req.body.rideData[k].accelerometerY,
                accelerometerY: req.body.rideData[k].accelerometerZ,
                gyroscopeX: req.body.rideData[k].gyroscopeX,
                gyroscopeY: req.body.rideData[k].gyroscopeY,
                gyroscopeZ: req.body.rideData[k].gyroscopeZ
            }));
        }

        // Insert data to collection
        RideData.collection.insertMany(allRideData, function(err) {
            if (err) {
                res.status(400).json(err);
            }
        });
    } catch (err) {
        res.status(400).json(err);
    }

    // Evaluate ride and update rideRecord document with ride score using TrailBrake Judge service        
    const judgeApiUrl = `${process.env.JUDGE_URL}/rideScore`;
    const judgeApiOptions = {
        json: {
            rideData: allRideData
        },
        retry: {
            limit: 0
        }
    };
    
    try {
        const { rideScore, rideMeta } = await got.post(judgeApiUrl, judgeApiOptions).json();

        // Update document object properties
        rideRecord.rideScore = rideScore;
        rideRecord.rideMeta = rideMeta;
    } catch (err) {
        res.status(err.response.statusCode).json(JSON.parse(err.response.body));
    }

    rideRecord.save(function(err, doc) {
        if (err) {
            res.status(400).json(err);
        } else {
            res.status(201).json({
                message: 'New ride successfully created',
                rideRecord: doc
            });
        }
    });
});

// Delete one
router.delete('/:rideObjID', findRideRecordByID, async (req, res) => {
    try {
        RideData.deleteMany({ 
            'metadata.rideRecordID': res.rideRecord._id
        }, function(err) {
            if (err) {
                res.status(400).json(err);
            } 
        });
        await res.rideRecord.remove();

        res.json({ message: 'Successfully deleted ride' });
    } catch (err) {
        res.status(500).json({ 
            method: "router.delete()", 
            message: err.message 
        });
    }
});

export default router;

File: api/model/rides.go
package model

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

// Struct for MongoDB rideRecords document mapping
type RideRecord struct {
	ID        primitive.ObjectID `bson:"_id" json:"_id"`
	UserID    primitive.ObjectID `bson:"userId" json:"userId"`
	RideName  string             `bson:"rideName" json:"rideName"`
	RideDate  time.Time          `bson:"rideDate" json:"rideDate"`
	RideMeta  RideMeta           `bson:"rideMeta,OmitEmpty" json:"rideMeta,omitempty"`
	RideScore RideScore          `bson:"rideScore,OmitEmpty" json:"rideScore,omitempty"`
}

// Struct for MongoDB rideData document mapping
type RideDatum struct {
	Timestamp      time.Time          `bson:"timestamp" json:"timestamp"`
	RideRecordID   primitive.ObjectID `bson:"rideRecordId" json:"rideRecordId"`
	LocationLat    float64            `bson:"locationLat" json:"locationLat"`
	LocationLong   float64            `bson:"locationLong" json:"locationLong"`
	AccelerometerX float64            `bson:"accelerometerX" json:"accelerometerX"`
	AccelerometerY float64            `bson:"accelerometerY" json:"accelerometerY"`
	AccelerometerZ float64            `bson:"accelerometerZ" json:"accelerometerZ"`
	GyroscopeX     float64            `bson:"gyroscopeX" json:"gyroscopeX"`
	GyroscopeY     float64            `bson:"gyroscopeY" json:"gyroscopeY"`
	GyroscopeZ     float64            `bson:"gyroscopeZ" json:"gyroscopeZ"`
	RotationX      float64            `bson:"rotationX" json:"rotationX"`
	RotationY      float64            `bson:"rotationY" json:"rotationY"`
	RotationZ      float64            `bson:"rotationZ" json:"rotationZ"`
	RotationW      float64            `bson:"rotationW" json:"rotationW"`
}

// Struct for rideMeta field in rideRecords documents
type RideMeta struct {
	Distance        float64 `bson:"distance" json:"distance"`
	Duration        float64 `bson:"duration" json:"duration"`
	MaxAcceleration float64 `bson:"maxAcceleration" json:"maxAcceleration"`
	AccelerationRMS float64 `bson:"accelerationRms" json:"accelerationRms"`
}

// Struct for rideScore field in rideRecords documents
type RideScore struct {
	Overall      float64 `bson:"overall" json:"overall"`
	Speed        float64 `bson:"speed" json:"speed"`
	Comfort      float64 `bson:"comfort" json:"comfort"`
	Acceleration float64 `bson:"acceleration" json:"acceleration"`
	Braking      float64 `bson:"braking" json:"braking"`
	Cornering    float64 `bson:"cornering" json:"cornering"`
}

// Struct for JSON response at GET /rides/{rideId} endpoint
type Ride struct {
	ID        primitive.ObjectID `bson:"_id" json:"_id"`
	RideName  string             `bson:"rideName" json:"rideName"`
	RideDate  time.Time          `bson:"rideDate" json:"rideDate"`
	RideMeta  RideMeta           `bson:"rideMeta" json:"rideMeta"`
	RideScore RideScore          `bson:"rideScore" json:"rideScore"`
	RideData  []RideDatum        `bson:"rideData" json:"rideData"`
}

// Struct for Judge service response
type JudgeResult struct {
	RideScore RideScore `json:"rideScore"`
	RideMeta  RideMeta  `json:"rideMeta"`
}
